### PR TITLE
Add unit tests for HiveMerger

### PR DESCRIPTION
**Pull request by joelverhagen**
There are two approaches in this PR for unit testing the complex HiveMerger component (PR'd with https://github.com/NuGet/NuGet.Services.Metadata/pull/707).

1. Traditional unit tests covered happy path case and edge cases.
2. A "full enumeration" test that generates all possible subsets of a set of versions for both the existing versions in a registration index and the versions coming from the catalog.

The reason this full enumeration suite is valuable is because the naive implementation of HiveMerger is trivial and can be done in the assertion step of a unit test. The implementation we have for product code (not tests) is more complex because it needs to be more performance and avoid unnecessary HTTP requests.

Progress on https://github.com/NuGet/NuGetGallery/issues/7736

### CODE REVIEW

**Comment by joelverhagen** [comment]
:bell: could I get some eyes on this?

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.MergeAsync.cs [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+using System.Threading.Tasks;
+using NuGet.Versioning;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace NuGet.Jobs.Catalog2Registration
+{
+    public partial class HiveMergerFacts
+    {
+        public class MergeAsync : Facts
```
Comment (agr) [inline tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.MergeAsync.cs]: Please, don't name test classes exactly like methods, it messes up Ctrl+T search in VS: classes pop up in search higher than methods and in most cases you'd want method.

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.Support.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+            return versions;
+        }
+
+        private static List<CatalogCommitItem> MakeSortedCatalog(params VersionAction[] versions)
+        {
+            return MakeSortedCatalog(versions.ToList());
+        }
+
+        private static List<CatalogCommitItem> MakeSortedCatalog(List<VersionAction> sortedVersionActions)
```
Comment (agr) [inline tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.Support.cs]: `ICollection<VersionAction>`? Will save the `ToList()` call couple of lines above.

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+                // Build the input data structures.
+                var sortedCatalog = MakeSortedCatalog(allUpdated);
+                var indexInfo = MakeIndexInfo(existing, config.MaxLeavesPerPage, versionToNormalized);
+
+                // Determine which pages will definitely not be affected.
+                var unaffectedPages = new List<PageInfo>();
+                if (existing.Any())
+                {
+                    var minVersion = allUpdated.Min(x => x.Version);
+                    for (int i = 0; i < indexInfo.Items.Count - 1; i++)
```
Comment (agr) [inline tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs]: `unaffectedPages = indexInfo.Items.Where(ii => ii.Upper < minVersion).ToList()`?

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+                }
+
+                // Verify the modified pages are in the index.
+                foreach (var pageInfo in result.ModifiedPages)
+                {
+                    Assert.True(indexInfo.Items.Contains(pageInfo), "A modified page must be in the index.");
+                }
+
+                // Verify the leaf item infos match the leaf items.
+                Assert.True(
```
Comment (agr) [inline tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs]: `Assert.Equal`

**Comment by joelverhagen** [comment]
> I am a little bit concerned that all versions in all tests have minor and patch portions set to 0. Do we have tests that make sure that inside the `HiveMerger` correct version comparison is used?

Great point! I'll look into it.

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.MergeAsync.cs (hunk 1 above)
Comment (joelverhagen) [reply tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.MergeAsync.cs]: Sounds good. Would something like `TheMergeAsyncMethod` be better?

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs (hunk 4 above)
Comment (joelverhagen) [reply tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs]: During the debugging I found that the custom messages were more useful than the counts that come from `Assert.Equal`.

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.MergeAsync.cs (hunk 1 above)
Comment (joelverhagen) [reply tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.MergeAsync.cs]: Did that!

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.Support.cs (hunk 2 above)
Comment (joelverhagen) [reply tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.Support.cs]: Done.

File: tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs (hunk 3 above)
Comment (joelverhagen) [reply tests/NuGet.Jobs.Catalog2Registration.Tests/Hives/HiveMergerFacts.FullEnumeration.cs]: Done.

**Comment by joelverhagen** [comment]
Added more version variants.
